core/src/test/scala/com/databricks/labs/remorph/RootTableIdentifierTest.scala

```diff
@@ -0,0 +1,254 @@
+package com.databricks.labs.remorph
+
+import org.scalatest.funsuite.AnyFunSuite
+import com.databricks.labs.remorph.transpilers.{BaseTranspiler, SnowflakeToDatabricksTranspiler, TSqlToDatabricksTranspiler, Transpiler}
+import org.scalatest.prop.TableDrivenPropertyChecks
+
+class SnowflakeToDatabricksTranspilerTest extends RootTableIdentifierTest {
+  override def transpiler: Transpiler = new SnowflakeToDatabricksTranspiler()
+}
+
+class TSqlToDatabricksTranspilerTest extends RootTableIdentifierTest {
+  override def transpiler: Transpiler = new TSqlToDatabricksTranspiler()
+}
+
+trait RootTableIdentifierTest extends AnyFunSuite with TableDrivenPropertyChecks {
+
+  def transpiler: Transpiler
+
+  test("processQuery should correctly process SQL and update Lineage Insert Select") {
+    val sql = """INSERT INTO output_table AS SELECT col1, col2
+                |FROM input_table_1 JOIN input_table_2
+                |ON input_table_1.col1 = input_table_2.col1""".stripMargin
+    val dag = new Lineage()
+    val rootTableIdentifier = new RootTableIdentifier(transpiler)
+
+    val resultDAG = rootTableIdentifier.processQuery(sql, dag)
+    val expectedNodes = Set(Node("output_table"), Node("input_table_1"), Node("input_table_2"))
+    val expectedEdges =
+      Set(
+        Edge(Node("input_table_1"), Action.Write, Node("output_table")),
+        Edge(Node("input_table_2"), Action.Write, Node("output_table")))
+
+    assert(resultDAG.getNodes == expectedNodes)
+    assert(resultDAG.getEdges == expectedEdges)
+  }
```

Comment:
this class design is brittle, as it leaks graph to the consumer of the instance. Public API has to be:

```
rootDiscovery = RootTableIdentifier(parser, Seq(
  "SELECT ... query1",
  "CREATE TABLE query2 AS SELECT ...",
"CREATE TABLE query3 AS SELECT ...",
"CREATE TABLE query4 AS SELECT ...",
  "SELECT ... query5",
))

assert(rootDiscovery.roots == Seq("query2", "query87"))
assert(rootDiscovery.sorted == Seq("query2", "query87", "query89", "query2", "query1", "query243", ...))
```

current test cases are not good enough, as they focus on the wrong thing and leak too much implementaiton detail.

Reply:
Done